index.bs

```diff
@@ -2015,14 +1903,67 @@ Issue: This needs to be generalised to work with realms too.
 <div algorithm="remote end steps for session.unsubscribe">
 The [=remote end steps=] with |session| and |command parameters| are:
 
-1. Let the |list of event names| be the value of the <code>events</code> field of
-   |command parameters|.
+1. If |command parameters| matches the <code>session.SubscriptionRequest</code> production:
+
+   1. Let the |event names| be the value of the <code>events</code> field of
+      |command parameters|.
+
+   1. Let the |contexts| be the value of the <code>contexts</code>
+      field of |command parameters| if it is present or an empty [=/set=] if it isn't.
+
+   1. Let |new subscriptions| to be a [=/list=].
+
+   1. For each |subscription| of |session|'s [=subscriptions=]:
+
+      1. If [=set/intersection=] of |subscription|'s event names and |event names| is an empty [=/set=],
+         continue.
+
+      1. If |contexts| is an empty [=/set=]:
+
+        1. |=list/Remove=| all |event names| from |subscription|'s event names.
+
+        1. If |subscription|'s event names is not empty:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+      1. Otherwise:
+
+        1. If |subscription| is [=subscription/global=]:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+        1. Otherwise:
+
+          1. Let |event map| be an empty [=/map=].
+
+          1. For each |event name| in |subscription|'s event names:
+
+              1. Set |event map|[|event name|] to [=set/clone=] of |subscription|'s contexts.
+
+          1. For each |event name| in |event names|:
+
+              1. If |event map|[|event name|] does not exist,
+                 continue.
+
+              1. Remove |contexts| from |event map|[|event name|].
+
+              1. If |event map|[|event name|] is an empty [=/set=],
+
+                1. [=map/Remove=] |event map|[|event name|].
+
+          1. For each |event name| → |remaining contexts| in |event map|:
```

Comment:
So here we end up maximally splitting, so that each entry contains a single event name, and a list of contexts.

I don't think it's wrong, but I also don't think it's better than my suggestions :) (either we do this from the start and have a struct containing an id, a single event name, and a list of contexts, or we have a struct containing an id and a map from event names to contexts).

I won't block on that however.

Reply:
I still hope that we will be able to deprecate this branch. I do not think the current PR's structure is better for the use case of splitting subscriptions but if we pre-process subscriptions upfront like you suggest then we need to also duplicate context ids in the map even if we never split the subscription and take the structure into account in every algorithm. I could also add additional steps to group the events for the same context sets but I think it is also an option to leave this optimization to implementations as the number of actual subscription entries should not be observable (if we want to make it observable it is a different topic).

Reply:
I feel like the data structure where we have an id and a map from event names to what the subscriptions are works pretty well across the entire spec, but you're writing the patch so it's your decision.